sys-utils/rollup/rollup.cpp

```diff
@@ -265,97 +237,94 @@ static int write_voucher(void) try {
 static int throw_exception(void) try {
+    rollup r;
     auto ji = nlohmann::json::parse(read_input());
-    auto payload = ji["payload"].get<std::string>();
-    struct rollup_exception e;
-    memset(&e, 0, sizeof(e));
-    e.payload.data = reinterpret_cast<unsigned char *>(payload.data());
-    e.payload.length = payload.size();
-    file_desc_ioctl(unique_open(ROLLUP_DEVICE_NAME, O_RDWR), IOCTL_ROLLUP_THROW_EXCEPTION, &e);
-    return 0;
+    auto payload = unhex(ji["payload"].get<std::string>());
+    return cmt_rollup_emit_exception(r, payload.size(), reinterpret_cast<uint8_t *>(payload.data()));
 } catch (std::exception &x) {
     std::cerr << x.what() << '\n';
     return 1;
 }
 
 // Read advance state data from driver, write to output
-static void write_advance_state(const unique_file_desc &fd, const struct rollup_finish *f) {
-    struct rollup_advance_state r;
-    memset(&r, 0, sizeof(r));
-    resize_bytes(&r.payload, f->next_request_payload_length);
-    file_desc_ioctl(fd, IOCTL_ROLLUP_READ_ADVANCE_STATE, &r);
-    auto payload = std::string_view{reinterpret_cast<const char *>(r.payload.data), r.payload.length};
-    const auto &m = r.metadata;
+static void write_advance_state(rollup &r, cmt_rollup_finish_t *f) {
+    (void) f;
+    cmt_rollup_advance_t advance;
+    if (cmt_rollup_read_advance_state(r, &advance))
+        return;
+
     nlohmann::json j = {{"request_type", "advance_state"},
         {"data",
-            {{"payload", payload},
-                {"metadata",
-                    {{"msg_sender", hex(m.msg_sender, sizeof(m.msg_sender))}, {"epoch_index", m.epoch_index},
-                        {"input_index", m.input_index}, {"block_number", m.block_number},
-                        {"timestamp", m.timestamp}}}}}};
+            {
+                {"chain_id", advance.chain_id},
+                {"payload", hex(reinterpret_cast<const uint8_t *>(advance.payload), advance.payload_length)},
+                {"app_contract", hex(advance.app_contract, sizeof(advance.app_contract))},
+                {"msg_sender", hex(advance.msg_sender, sizeof(advance.msg_sender))},
+                {"block_number", advance.block_number},
+                {"block_timestamp", advance.block_timestamp},
+                {"index", advance.index},
+            }}};
     std::cout << j.dump(2) << '\n';
 }
 
 // Read inspect state data from driver, write to output
-static void write_inspect_state(const unique_file_desc &fd, const struct rollup_finish *f) {
-    struct rollup_inspect_state r;
-    memset(&r, 0, sizeof(r));
-    resize_bytes(&r.payload, f->next_request_payload_length);
-    file_desc_ioctl(fd, IOCTL_ROLLUP_READ_INSPECT_STATE, &r);
-    auto payload = std::string_view{reinterpret_cast<const char *>(r.payload.data), r.payload.length};
+static void write_inspect_state(rollup &r, cmt_rollup_finish_t *f) {
+    (void) f;
+    cmt_rollup_inspect_t inspect;
+    if (cmt_rollup_read_inspect_state(r, &inspect))
```

Comment:
Tool should fail, but is ignoring malformed inputs
(Inspect is not encoded anymore, but it may still have an invalid length)